File: contest/LeetCode_Weekly_Contest_206/Min_Cost_to_Connect_All_Points.cpp
class Solution {
public:
    struct Edge {
        int u, v, w;
    };
    
    int n = 0;
    Edge e[1000005];
    vector<int> p;
    
    int find(int u) {
        return u == p[u] ? u : p[u] = find(p[u]);
    }
    
    int minCostConnectPoints(vector<vector<int>>& points) {
        n = points.size();
        
        int k = 0;
        for (int i = 0; i < n; ++i)
            for (int j = i+1; j < n; ++j) {
                int xi = points[i][0], yi = points[i][1];
                int xj = points[j][0], yj = points[j][1];
                int dist = abs(xi - xj) + abs(yi - yj);
                e[k++] = Edge{i, j, dist};
            }
        
        sort(e, e + k, [](const Edge& a, const Edge& b) { 
            return a.w < b.w; 
        });
        
        p.resize(n);
        for (int i = 0; i < n; ++i) p[i] = i;
        
        int ret = 0;
        for (int i = 0; i < k; ++i) {
            int u = e[i].u,  v = e[i].v, w = e[i].w;
            int pu = find(u);
            int pv = find(v);
            if (pu == pv) continue;
            p[pu] = pv;
            ret += w;
        }
        return ret;
    }
};

File: contest/LeetCode_Weekly_Contest_206/Min_Cost_to_Connect_All_Points.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        dist = [10**6 * 4 + 1] * n
        visited = [False] * n

        s = 0
        dist[0] = 0
        visited[0] = True
        for i in range(n):
            next_s = -1
            for j in range(n):
                if not visited[j]:
                    dist[j] = min(dist[j], abs(points[s][0] - points[j][0]) + abs(points[s][1] - points[j][1]))
                    if next_s == -1 or dist[j] < dist[next_s]:
                        next_s = j
            s = next_s
            visited[s] = True
        return sum(dist)

